### PR TITLE
add a phase gradient regularization and speckle contrast regularization

### PR DESCRIPTION
**Pull request by zhurz21**
Hello Prof. Kaan Aksit,

     Last week we talk about phase gradient and speckle contrast in hologram generation.

     And I have written those two regularization terms in torch. I placed the python file under odak/odak/learn/perception/phase_gradient_regularization.py. And maybe you can test if it can work well.

Please don't be hesitate to connect with me if you have any problems.

Looking forward to your email.

Best regards,

Runze Zhu, Tsinghua University, Beijing, China

### CODE REVIEW

**Comment by kunguz** [comment]
@zhurz21 Thank you for making this happen! a few minor changes are required before I give this one try and merge it to our repository:

- Please drop using `numpy` and `math` in `phase_gradient_regularization.py`. Keep it all `torch`.
- Regarding `docstrings`, please follow the same style in our other files.
- Move the `phase_gradient_regularization.py` to `odak/learn/wave/loss.py`.
- Please add `docstrings` to `forward` and `functional_conv2d` functions.
- Please add `from .loss import *` to `odak/learn/wave/__init__.py`.

Once we have this code in our repository and verify it to be helpful for our project. We should also plan on adding documentation on this to [here](https://kunguz.github.io/odak/cgh/). Note that our documentation is within the same repository living in `docs` folder as `md` files.

**Comment by zhurz21** [comment]
@kunguz Hello Prof. Kaan Akşit, thanks for your reply, and I have standardized the code and docstrings.

As it's my first time making a contribution to an open-source project, some of my operations may be inappropriate, but I'm glad to study, so please don't be hesitate to correct me.

Please inform me if there is a need for writing documentation after your testing.

**Comment by kunguz** [comment]
Excellent! This commit looks all good to me—one more thing before we ship this to @KorayKavakli  so that he can try. We need an example.

In the `test` folder, you will find some unit tests for other cases. So, please add a new one as `test/test_learn_loss.py`.

In this one, generate a random field by 

``` Python
from odak.learn.wave import generate_complex_field
import torch
import sys

def main():
    torch.random.seed()
    phase = torch.rand(1920,1080)
    amplitude = torch.rand_like(phase)
    complex_field = generate_complex_field(amplitude, phase)
    # Add here your example for the loss function usage.
    assert True == True

if  __name__ ==  '__main__':
    sys.exit(main())
```

Note that we initiate these unit tests at every commit on GitHub servers. We have to make sure that it passes the test to state the code is reliable. To test unit tests locally, make sure that you have `pytest` installed:

```
pip3 install pytest
```

Make sure to test all tests by running `pytest` in the root folder. 


In case, you are interested in learning more about our ways of developing things for this library, then I can point you to [our documentation](https://kunguz.github.io/odak/contributing/) for greater details.

**Comment by zhurz21** [comment]
Hello professor Kaan Akşit, I have added a test file as `test/test_learn_loss.py` and tested it using 'pytest' locally. I think the code is reliable to use now.

**Comment by kunguz** [comment]
@zhurz21  I merged the final version into the main branch. At this stage, @KorayKavakli will test your loss function in his codebase. Please make sure to pass me your full name used in your publications and your [ORCID account](https://orcid.org/) so that I can add your name to `CITATION.cff` and `THANKS.txt`.
